flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/WatermarkGeneratorCodeGenerator.scala

```diff
@@ -0,0 +1,91 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.planner.codegen
+
+import org.apache.flink.configuration.Configuration
+import org.apache.flink.table.api.TableConfig
+import org.apache.flink.table.planner.calcite.FlinkTypeFactory
+import org.apache.flink.table.planner.codegen.CodeGenUtils.{BASE_ROW, newName}
+import org.apache.flink.table.planner.codegen.Indenter.toISC
+import org.apache.flink.table.runtime.generated.GeneratedWatermarkGenerator
+import org.apache.flink.table.runtime.operators.wmassigners.WatermarkGenerator
+import org.apache.flink.table.types.logical.{RowType, TimestampType}
+
+import org.apache.calcite.rex.RexNode
+
+/**
+  * A code generator for generating [[WatermarkGenerator]]s.
+  */
+object WatermarkGeneratorCodeGenerator {
+
+  def generateWatermarkGenerator(
+      config: TableConfig,
+      inputType: RowType,
+      watermarkExpr: RexNode): GeneratedWatermarkGenerator = {
+    // validation
+    val watermarkOutputType = FlinkTypeFactory.toLogicalType(watermarkExpr.getType)
+    if (!watermarkOutputType.equals(new TimestampType(3))) {
```

Comment:
Relax this? IMO if `precision >= 3` is enough

Reply:
I agree we can support `precision >= 3` but that requires the data format (BaseRow) should support high precisions. Until then, we can update the code generation logic. 
But for now, we can' support high precision. This is a safer protection.

Reply:
I recall my first comment. I think as long as the type is `TimestampType` is enough, we shouldn't force the precision to be 3.

Reply:
But for the runtime, we only support precision 3, this is similar to the rowtime type. Maybe we can add validation to the earlier stage (e.g. TableSchema) to restrict this precision.

Reply:
With precision less than 3, we still can get millisecond from the field. For example, if the precision is 0, the millisecond we get will always end up in `000`

Reply:
I got your point. I will change this to `if (!(timestamp type && precision <= 3)) { exception }`.

Reply:
I will just check timestamp type here.